### Содержание: 
- [Лабораторная работа № 2](#Lab2)
- [Лабораторная работа № 3](#Lab3)
- [Лабораторная работа № 4](#Lab4)
- [Лабораторная работа № 5](#Lab5)
- [Лабораторная работа № 6](#Lab6)
- [Лабораторная работа № 7](#Lab7)

## Лабораторная работа № 2 <a name="Lab2"></a>

* Выбрать и скачать на сайте Kaggle.com набор данных (DataSet) в формате csv, содержащий не менее 3-х категориальных атрибутов; 
* Ссылку на выбранный DataSet скинуть в чат группы. 
* DataSet в группе не должны повторяться. 
* Создать notebook на Colab и подключить к нему выбранный DataSet.


In [269]:
import pandas as pd
import numpy as np
datafile = open("../data/athlete_events.csv", "r")
df = pd.read_csv(datafile)
df

ID                      Name Sex   Age  Height  Weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

                  Team  NOC        Games  Year  Season            City  \
0                China  CHN  1992 Summer  1992  Summer       Barcelona   
1                China  CHN  2012 Summer  2012  Summer          London   
2              Denmark  DEN  1920 Summer  1920  Summer       Antwerpen   
3       Denmark/Sweden  DEN  1900 Summer  1900  Summer           Paris   
4          Netherlands  NED  1988 Winter  1988  Winter         Calgary   
...                ...  ...          ...   ...     ...             ...   
271111        Poland-1  POL  1976 Winter  1976  Winter       Innsbruck   
271112          Poland  POL  2014 Winter  2014  Winter           Sochi   
271113          Poland  POL  2014 Winter  2014  Winter           Sochi   
271114          Poland  POL  1998 Winter  1998  Winter          Nagano   
271115          Poland  POL  2002 Winter  2002  Winter  Salt Lake City   

                Sport                                     Event Medal  
0          Basketball               Basketball Men's Basketball   NaN  
1                Judo              Judo Men's Extra-Lightweight   NaN  
2            Football                   Football Men's Football   NaN  
3          Tug-Of-War               Tug-Of-War Men's Tug-Of-War  Gold  
4       Speed Skating          Speed Skating Women's 500 metres   NaN  
...               ...                                       ...   ...  
271111           Luge                Luge Mixed (Men)'s Doubles   NaN  
271112    Ski Jumping  Ski Jumping Men's Large Hill, Individual   NaN  
271113    Ski Jumping        Ski Jumping Men's Large Hill, Team   NaN  
271114      Bobsleigh                      Bobsleigh Men's Four   NaN  
271115      Bobsleigh                      Bobsleigh Men's Four   NaN  

[271116 rows x 15 columns]

## Лабораторная работа № 3. Поиск функциональных зависимостей <a name="Lr3"></a>

- Написать на языке python обнаружения функциональных зависимостей в DataSet библиотеки pandas

In [270]:
from itertools import permutations

def check_fd(series_1, series_2):
    if series_1.nunique() < series_2.nunique():
#         print("Violates by unique")
        return False

    values = {}
    for val_1, val_2 in zip(series_1, series_2):
        if val_1 not in values:
            values[val_1] = val_2
        elif values[val_1] != val_2:
#             print(f"Violates by values: {values[val_1]}, {val_2}")
            return False

    return True

for attr_1, attr_2 in permutations(df.columns, 2):
#     print(f"Check {attr_1} and {attr_2}")
    if check_fd(df[attr_1], df[attr_2]):
        print(f"{attr_1} -> {attr_2}")
#     print("CHECKED-------------\n\n")

ID -> Name
ID -> Sex
Games -> Year
Games -> Season
City -> Season
Event -> Sport


## Лабораторная работа № 4 <a name="Lab4"></a>

- Проанализировать типы атрибутов DataSet, определить тип (номинальный, порядковый, дискретный, (интервальный), непрерывный), домен (множество принимаемых значений) и шкалу;
- Выявить и устранить аномалии в DataSet (неатомарные атрибуты, ошибки ввода, гетерогенные данные, ошибка определения типа, подмена сущностей экземплярами);
- Сохранить предобработанный DataSet и в дальнейшем работать с ним.

### Типы данных:

ID - Атрибутивный,порядковый.	
1 - 135571

Name - Атрибутивный,номинальный.

Sex - Атрибутивный,номинальный.
M, F

Age - Переменный,относительный.
10 - 97

Height - Переменный,относительный.
127 - 226

Weight - Переменный,относительный.
25 - 214

Team - Атрибутивный,номинальный.

NOC - Атрибутивный,номинальный.

Games - Атрибутивный,порядковый.
1896 Summer - 2016 Summer

Year - Переменный,интервальный.
1896 - 2016

Season - Атрибутивный,номинальный.
Winter - Summer

City - Атрибутивный,номинальный.

Sport - Атрибутивный,номинальный.

Event - Атрибутивный,номинальный.

Medal - Атрибутивный,порядковый.
Gold,Silver,Bronze,None

### Устранение аномалий и рефакторинг датасета

#### Удалим столбец Games, дублирующий данные из колонок Year и Season

In [271]:
df = df.drop("Games", axis=1)
df

ID                      Name Sex   Age  Height  Weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

                  Team  NOC  Year  Season            City          Sport  \
0                China  CHN  1992  Summer       Barcelona     Basketball   
1                China  CHN  2012  Summer          London           Judo   
2              Denmark  DEN  1920  Summer       Antwerpen       Football   
3       Denmark/Sweden  DEN  1900  Summer           Paris     Tug-Of-War   
4          Netherlands  NED  1988  Winter         Calgary  Speed Skating   
...                ...  ...   ...     ...             ...            ...   
271111        Poland-1  POL  1976  Winter       Innsbruck           Luge   
271112          Poland  POL  2014  Winter           Sochi    Ski Jumping   
271113          Poland  POL  2014  Winter           Sochi    Ski Jumping   
271114          Poland  POL  1998  Winter          Nagano      Bobsleigh   
271115          Poland  POL  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal  
0                    Basketball Men's Basketball   NaN  
1                   Judo Men's Extra-Lightweight   NaN  
2                        Football Men's Football   NaN  
3                    Tug-Of-War Men's Tug-Of-War  Gold  
4               Speed Skating Women's 500 metres   NaN  
...                                          ...   ...  
271111                Luge Mixed (Men)'s Doubles   NaN  
271112  Ski Jumping Men's Large Hill, Individual   NaN  
271113        Ski Jumping Men's Large Hill, Team   NaN  
271114                      Bobsleigh Men's Four   NaN  
271115                      Bobsleigh Men's Four   NaN  

[271116 rows x 14 columns]

#### Исправим ошибки ввода в столбце полного названия страны на основании наиболее частого сочетания (NOC, Country)

In [272]:
from collections import Counter

most_common_team_name = df.groupby("NOC").apply(lambda g: Counter(g.Team)).apply(
    lambda g: max(g.items(), key=lambda pair: pair[1])[0]
)
fixed_country_names = df.NOC.map(lambda noc: most_common_team_name[noc])
df.Team = fixed_country_names
df

ID                      Name Sex   Age  Height  Weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

               Team  NOC  Year  Season            City          Sport  \
0             China  CHN  1992  Summer       Barcelona     Basketball   
1             China  CHN  2012  Summer          London           Judo   
2           Denmark  DEN  1920  Summer       Antwerpen       Football   
3           Denmark  DEN  1900  Summer           Paris     Tug-Of-War   
4       Netherlands  NED  1988  Winter         Calgary  Speed Skating   
...             ...  ...   ...     ...             ...            ...   
271111       Poland  POL  1976  Winter       Innsbruck           Luge   
271112       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271113       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271114       Poland  POL  1998  Winter          Nagano      Bobsleigh   
271115       Poland  POL  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal  
0                    Basketball Men's Basketball   NaN  
1                   Judo Men's Extra-Lightweight   NaN  
2                        Football Men's Football   NaN  
3                    Tug-Of-War Men's Tug-Of-War  Gold  
4               Speed Skating Women's 500 metres   NaN  
...                                          ...   ...  
271111                Luge Mixed (Men)'s Doubles   NaN  
271112  Ski Jumping Men's Large Hill, Individual   NaN  
271113        Ski Jumping Men's Large Hill, Team   NaN  
271114                      Bobsleigh Men's Four   NaN  
271115                      Bobsleigh Men's Four   NaN  

[271116 rows x 14 columns]

#### Установим значение "None" вместо NaN в колонке медалей:

In [273]:
df.Medal = df.Medal.map(lambda med: None if pd.isnull(med) else med)
df

ID                      Name Sex   Age  Height  Weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

               Team  NOC  Year  Season            City          Sport  \
0             China  CHN  1992  Summer       Barcelona     Basketball   
1             China  CHN  2012  Summer          London           Judo   
2           Denmark  DEN  1920  Summer       Antwerpen       Football   
3           Denmark  DEN  1900  Summer           Paris     Tug-Of-War   
4       Netherlands  NED  1988  Winter         Calgary  Speed Skating   
...             ...  ...   ...     ...             ...            ...   
271111       Poland  POL  1976  Winter       Innsbruck           Luge   
271112       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271113       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271114       Poland  POL  1998  Winter          Nagano      Bobsleigh   
271115       Poland  POL  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal  
0                    Basketball Men's Basketball  None  
1                   Judo Men's Extra-Lightweight  None  
2                        Football Men's Football  None  
3                    Tug-Of-War Men's Tug-Of-War  Gold  
4               Speed Skating Women's 500 metres  None  
...                                          ...   ...  
271111                Luge Mixed (Men)'s Doubles  None  
271112  Ski Jumping Men's Large Hill, Individual  None  
271113        Ski Jumping Men's Large Hill, Team  None  
271114                      Bobsleigh Men's Four  None  
271115                      Bobsleigh Men's Four  None  

[271116 rows x 14 columns]

#### Добавим отсутствующие значения веса и роста путем вычисления среднего арифметического

In [268]:
means = df.groupby(["Sex", "Event"]).mean()
mean_heights = means["Height"]
mean_weights = means["Weight"]

aliases = df[["Sex", "Event"]].apply(lambda row: (row.Sex, row.Event), axis=1)
weights, heights = df.Weight, df.Height

weight_isnull = weights.map(pd.isnull)
height_isnull = heights.map(pd.isnull)

normalized_weihgts_asindex = df.index.map(lambda i: mean_weights[aliases[i]] if weight_isnull[i] else weights[i])
normalized_weights = pd.Series(normalized_weihgts_asindex.values)

normalized_heights_asindex = df.index.map(lambda i: mean_heights[aliases[i]] if height_isnull[i] else heights[i])
normalized_heights = pd.Series(normalized_heights_asindex.values)

df.Height = normalized_heights
df.Weight = normalized_weights

df

ID                      Name Sex   Age      Height     Weight  \
0            1                 A Dijiang   M  24.0  180.000000  80.000000   
1            2                  A Lamusi   M  23.0  170.000000  60.000000   
2            3       Gunnar Nielsen Aaby   M  24.0  177.480339  73.086644   
3            4      Edgar Lindenau Aabye   M  34.0  182.480000  95.615385   
4            5  Christine Jacoba Aaftink   F  21.0  185.000000  82.000000   
...        ...                       ...  ..   ...         ...        ...   
271111  135569                Andrzej ya   M  29.0  179.000000  89.000000   
271112  135570                  Piotr ya   M  27.0  176.000000  59.000000   
271113  135570                  Piotr ya   M  27.0  176.000000  59.000000   
271114  135571        Tomasz Ireneusz ya   M  30.0  185.000000  96.000000   
271115  135571        Tomasz Ireneusz ya   M  34.0  185.000000  96.000000   

               Team  NOC  Year  Season            City          Sport  \
0             China  CHN  1992  Summer       Barcelona     Basketball   
1             China  CHN  2012  Summer          London           Judo   
2           Denmark  DEN  1920  Summer       Antwerpen       Football   
3           Denmark  DEN  1900  Summer           Paris     Tug-Of-War   
4       Netherlands  NED  1988  Winter         Calgary  Speed Skating   
...             ...  ...   ...     ...             ...            ...   
271111       Poland  POL  1976  Winter       Innsbruck           Luge   
271112       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271113       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271114       Poland  POL  1998  Winter          Nagano      Bobsleigh   
271115       Poland  POL  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal  
0                    Basketball Men's Basketball  None  
1                   Judo Men's Extra-Lightweight  None  
2                        Football Men's Football  None  
3                    Tug-Of-War Men's Tug-Of-War  Gold  
4               Speed Skating Women's 500 metres  None  
...                                          ...   ...  
271111                Luge Mixed (Men)'s Doubles  None  
271112  Ski Jumping Men's Large Hill, Individual  None  
271113        Ski Jumping Men's Large Hill, Team  None  
271114                      Bobsleigh Men's Four  None  
271115                      Bobsleigh Men's Four  None  

[271116 rows x 14 columns]

ID                      Name Sex   Age      Height     Weight  \
0            1                 A Dijiang   M  24.0  180.000000  80.000000   
1            2                  A Lamusi   M  23.0  170.000000  60.000000   
2            3       Gunnar Nielsen Aaby   M  24.0  177.480339  73.086644   
3            4      Edgar Lindenau Aabye   M  34.0  182.480000  95.615385   
4            5  Christine Jacoba Aaftink   F  21.0  185.000000  82.000000   
...        ...                       ...  ..   ...         ...        ...   
271111  135569                Andrzej ya   M  29.0  179.000000  89.000000   
271112  135570                  Piotr ya   M  27.0  176.000000  59.000000   
271113  135570                  Piotr ya   M  27.0  176.000000  59.000000   
271114  135571        Tomasz Ireneusz ya   M  30.0  185.000000  96.000000   
271115  135571        Tomasz Ireneusz ya   M  34.0  185.000000  96.000000   

               Team  NOC  Year  Season            City          Sport  \
0             China  CHN  1992  Summer       Barcelona     Basketball   
1             China  CHN  2012  Summer          London           Judo   
2           Denmark  DEN  1920  Summer       Antwerpen       Football   
3           Denmark  DEN  1900  Summer           Paris     Tug-Of-War   
4       Netherlands  NED  1988  Winter         Calgary  Speed Skating   
...             ...  ...   ...     ...             ...            ...   
271111       Poland  POL  1976  Winter       Innsbruck           Luge   
271112       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271113       Poland  POL  2014  Winter           Sochi    Ski Jumping   
271114       Poland  POL  1998  Winter          Nagano      Bobsleigh   
271115       Poland  POL  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal  
0                    Basketball Men's Basketball  None  
1                   Judo Men's Extra-Lightweight  None  
2                        Football Men's Football  None  
3                    Tug-Of-War Men's Tug-Of-War  Gold  
4               Speed Skating Women's 500 metres  None  
...                                          ...   ...  
271111                Luge Mixed (Men)'s Doubles  None  
271112  Ski Jumping Men's Large Hill, Individual  None  
271113        Ski Jumping Men's Large Hill, Team  None  
271114                      Bobsleigh Men's Four  None  
271115                      Bobsleigh Men's Four  None  

[271116 rows x 14 columns]

## Лабораторная работа № 5 <a name="Lr5"></a>

- Для предобработанного DataSet:
 - выявить ФЗ и построить ER-диаграмму;
 - построить нормализованную схему (3 НФ);
 - построить схему в нотации UML.

#### Выведем фукнциональные зависимости:

In [21]:
for attr_1, attr_2 in permutations(df.columns, 2):
    if check_fd(df[attr_1], df[attr_2]):
        print(f"{attr_1} -> {attr_2}")

ID -> Name
ID -> Sex
Team -> NOC
NOC -> Team
City -> Season
Event -> Sport


## Лабораторная работа № 6 <a name="Lr6"></a>

Создать таблицы в бд, заполнить данными из датасета

In [285]:
from sqlalchemy import *

engine = create_engine('postgresql://publications:publications1209@localhost:5432/refactoring')
metadata = MetaData(bind=engine)

metadata.drop_all()

In [ ]:

cond_table = Table('Condd', metadata,
                    Column('ID_condition',Integer,primary_key=True),
                    Column('Average_Cost_for_two',Float),
                    Column('Has_Table_booking',String(3)),
                    Column('Is_delivering_now',String(3)),
                    Column('Has_Online_delivery',String(3)),
                    Column('Switch_to_order_menu',String(3))
)
m_table = Table('Mm', metadata,
                    Column('ID_mark',Integer,primary_key=True),
                    Column('Votes',Float),
                    Column('Aggregate_rating',Float),
                    Column('Rating_color',String(20)),
                    Column('Rating_text',String(20))
)
cur_table = Table('Curr', metadata,
                    Column('ID_currency',Integer,primary_key=True),
                    Column('Currency',String(30)),
                    Column('Abbr_currency',String(20))
)
local_table = Table('Locall', metadata,
                    Column('ID_locality',Integer,primary_key=True),
                    Column('Country_Code',String(30)),
                    Column('Longitude',Float),
                    Column('Latitude',Float),
                    Column('Place',String(40)),
                    Column('Place_Verbose',String(80))
)
rest_cuis_table = Table('Res_Cuiss', metadata,
                    Column('ID_restoraunt',Integer,ForeignKey('Restauu.ID_restaurant'),nullable=False),
                    Column('ID_cuisine',Integer,primary_key=True)
)
cuis_table = Table('Cuisinee', metadata,
                    Column('ID_cuisine',Integer,ForeignKey('Res_Cuiss.ID_cuisine'),nullable=False),
                    Column('Cuisine_name',String(70))
)
restau_table = Table('Restauu', metadata,
                    Column('ID_restaurant',Integer,primary_key=True),
                    Column('Restaurant_name',String(100)),
                    Column('ID_locality',Integer,ForeignKey('Locall.ID_locality'), nullable=False),
                    Column('ID_condition',Integer,ForeignKey('Condd.ID_condition'),nullable=False),
                    Column('ID_mark',Integer,ForeignKey('Mm.ID_mark'),nullable=False),
                    Column('ID_currency',Integer,ForeignKey('Curr.ID_currency'),nullable=False)
)
metadata.create_all(engine)

cuisine_table

##Удаление всей БД
metadata.drop_all(bind=engine)
#for tbl in reversed(metadata.sorted_tables):
#    engine.execute(tbl.delete())

##Удаление таблицы
sql = text('DROP TABLE IF EXISTS currency_table;')
result = engine.execute(sql)